### PR TITLE
[processor/transform] Convert between sum and gauge in metric context

### LINKED ISSUE #20773: [processor/transform] Metric conversion functions use datapoint context instead of metric context
### Component(s)

processor/transform

### Describe the issue you're reporting

The function `convert_gauge_to_sum`, `convert_sum_to_gauge`, `convert_summary_sum_val_to_sum` and `convert_summary_count_val_to_sum` all are using the `datapoint` context, but I'm not sure I understand why. 

They all operate on the top-level "metric", so I think they should be using the `metric` context, since they should each run only once for each metric, and not once for each datapoint.

It seems like this was just retaining the old behavior before contexts were introduced, but I think these functions are actually better suited on the `metric` context.

### PR DESCRIPTION
**Pull request by graphaelli**
**Description:** Allow running OTTL `convert_sum_to_gauge` and `convert_gauge_to_sum` in metric context instead of datapoint when `processor.transform.ConvertBetweenSumAndGaugeMetricContext` is enabled

closes #20773

This is the result of an effort at contribfest at Kubecon NA '23.

### CODE REVIEW

**Comment by graphaelli** [comment]
Just realized drafts don't request reviews.  Applied the same changes for`convert_gauge_to_sum`, renamed the feature gate and added a changelog entry - ready for review now.

File: CHANGELOG.md
Comment (evan-bradley) [inline CHANGELOG.md:1]: Thanks for adding a changelog message. This file is autogenerated, use `make chlog-new` to generate a new changelog entry file, and edit that file to create the entry.

File: processor/transformprocessor/internal/metrics/functions_test.go [hunk 1]
```diff
@@ -31,6 +31,8 @@ func Test_DataPointFunctions(t *testing.T) {
 
 func Test_MetricFunctions(t *testing.T) {
```
Comment (evan-bradley) [inline processor/transformprocessor/internal/metrics/functions_test.go:32]: I think you'll need to toggle the feature gate in this test.

**Review by evan-bradley — COMMENTED** [review]
Thanks for contributing this! Everything mostly looks good to me.

**Comment by graphaelli** [comment]
not sure what happened there, thought I just merged main back in.

File: processor/transformprocessor/internal/metrics/functions_test.go (hunk 1 above)
Comment (graphaelli) [reply processor/transformprocessor/internal/metrics/functions_test.go:33]: thanks, done

File: CHANGELOG.md
Comment (graphaelli) [reply CHANGELOG.md:1]: Thanks, done.

**Comment by graphaelli** [comment]
@evan-bradley thanks for the first pass.  Addressed all comments and ready for a review.

**Review by TylerHelmuth — APPROVED** [review]

**Review by evan-bradley — APPROVED** [review]

**Comment by evan-bradley** [comment]
@graphaelli Sorry, looks like the go.mod changed. Could you rebase one more time?

**Pull request merged**